File: 자료구조/C++/백준-스택-2841.cpp
# include <iostream>
# include <stack>
# include <vector>
using namespace std;

int main()
{
	ios::sync_with_stdio(false);
	cin.tie(NULL);

	int repeat, plat;
	int num, plat_num;
	cin >> repeat >> plat;
	
	int result = 0;
	vector<stack<int>> melody(7);

	for (int i = 0; i < repeat; i++) 
	{
		cin >> num >> plat_num;

		if (!melody[num].empty() && melody[num].top() > plat_num)
		{
			while (!melody[num].empty() && melody[num].top() > plat_num) {
				melody[num].pop();
				result++;
			}
		}

		if (melody[num].empty() || melody[num].top() < plat_num)
		{
			result++;
			melody[num].push(plat_num);
		}
	}

	cout << result << endl;
	return 0;
}

File: 자료구조/Python/백준-스택-2841.py
import sys

def main():
    N, plat_N = map(int, input().split())

    stack = [[] for _ in range(7)]
    total = 0

    for i in range(N):
        num, plat = map(int, sys.stdin.readline().split())
        if not stack[num]:
            stack[num].append(plat)
            total += 1
        else:
            while stack[num] and plat < stack[num][-1]:
                stack[num].pop()
                total += 1
            if not stack[num] or plat > stack[num][-1]:
                stack[num].append(plat)
                total += 1
            else:
                continue
    print(total)


if __name__ == "__main__":
    main()